design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+    `behavior.yaml` file and automatically using the templating controller image
+    (configured with the `behavior.yaml`) when one was not supplied in the
+    package. Composition enables this without having to install a new
+    controller. The composition controllers run as part of core Crossplane and
+    dynamically reconcile new types that are created in response to the creation
+    of an `InfrastructureDefinition` / `InfrastructurePublication`.
+
+Because composition serves the same purposes as these existing packaging
+primitives, the current packaging system can be narrowed in scope while also
+supporting a similar set of use-cases. The immediate goals of this refactor are:
+
+- Deprecation and subsequent removal of the `PackageInstall` and
+  `StackDefinition` types.
+- Standardization on a single `Package` type as a unit of installation.
+- Support for upgrading a `Package` that is installed and running in a cluster.
+- Support for native installation of composition types, including
+  `InfrastructureDefinition`, `InfrastructurePublication`, and `Composition`.
+
+The refactor will involve abandoning the following goals of the current model:
+
+- Supporting running multiple versions of the same package in a cluster at the
+  same time. This includes installing the same CRDs from multiple packages.
+
+## Proposal
+
+This proposal broadly encompasses the following components:
+
+- `Package`: a cluster-scoped CRD that represents the existence of a package in
+  a cluster. Users create / update / delete `Package` instances to manage the
+  packages present in their cluster.
+- `PackageRevision`: a cluster-scoped CRD that represents a version of a package
+  that may or may not be active in a cluster. Many `PackageRevision` instances
+  may exist for a single `Package`, but only one can be active at a given time.
+  `PackageRevision` instances are named the `sha256` hash of the image they are
+  represent.
+- _`Package` controller_: a controller responsible for observing events on
+  `Package` instances, creating new `PackageRevision` instances, and managing
+  which revision is active.
+- _`PackageRevision` controller_: a controller responsible for observing events
+  on `PackageRevision` instances, installing new CRDs /
+  `InfrastructureDefinition` / `InfrastructurePublication` / `Composition`
+  objects, starting package controllers, and cleaning up these objects when a
+  package is upgraded.
+
+These types and controllers overlap in functionality with some of the existing
+types and controllers that are currently present in the package manager
+ecosystem. However, their implementation should not immediately break any of the
+current features supported by the package manager. Removal of
+`ClusterPackageInstall`, `PackageInstall`, and `StackDefinition`, as well as
+refactor of `Package` and all controllers should follow a deprecation schedule
+that allows users to move to the new format.
```

Comment:
I can see that we'll need 2 different CRDs for resolving but I'm kind of suspicious about if we should have both of them now. I can see a pattern where we have only the lower level CRD (old `Package`, new `PackageRevision`) with late-inited fields, i.e. you'd supply image and the rest would be filled as it's installed. Then when we do the actual implementation of the resolver, we'd have a low level, flexible building block concept that we can construct the higher level CRD (old `PackageInstall`, new `Package`) with the actual code that does the resolving operation instead of writing it now and hope it's future-proof.

Reply:
This is related to my comment questioning feasibility of roll-back, which is the value prop of `PackageRevision`.

Reply:
@muvaf I think I would prefer to have a consistent interface for the user to interact with (i.e. the `Package`) rather than expose the `PackageRevision` and then build an abstraction on top of it. Regardless of resolving mechanisms, I don't think the `Package` type will change much from being a type where a user can specify an image and any other relevant information for getting that image (e.g. `imagePullSecrets`). It feels much more likely that the `PackageRevision` could be adjusted over time, but if we start out with the `Package` in front of that then we can avoid breaking users. Additionally, I don't love the late init pattern because we don't actually intend for a user to ever set most of those field, and would probably overwrite them if they did.

Reply:
> I think I would prefer to have a consistent interface for the user to interact with (i.e. the Package) rather than expose the PackageRevision and then build an abstraction on top of it. Regardless of resolving mechanisms, I don't think the Package type will change much from being a type where a user can specify an image and any other relevant information for getting that image (e.g. imagePullSecrets). It feels much more likely that the PackageRevision could be adjusted over time, but if we start out with the Package in front of that then we can avoid breaking users.

Do you think we will have another higher level type for resolution? Something like `Module -> N Package -> N PackageRevision`? Or are you imagining enhancing `Package` type to cover CRD resolution, too?

What I have in mind is `Module -> Package` where we implement the `Package` now with this design. `Package` would become something like:
```yaml
apiVersion: packages.crossplane.io/v1alpha1
kind: Package
metadata:
  name: provider-gcp
spec:
  package: crossplane/provider-gcp:v0.10.0
  uiMetadataRef:
    name: cfgmap
    namespace: crossplane-system
  imagePullSecretRef:
    name: secret
    namespace: crossplane-system
  envVars:
  ...
status:
  types:
    - rdsinstance.database.aws.crossplane.io/v1beta1
    - vpc.ec2.aws.crossplane.io/v1beta1
    ...
  conditions:
    ...
```

Then in the future, `Module` can built on top of this via a discovery mechanism like go.mod that associates the `spec.package` with the types in `status.types`. Users interact either with `Package` or `Module`, so, what they interact with doesn't really change when we implement the CRD resolution.

Reply:
@muvaf I do think we need a `Module` type to store the overall state of packages in the cluster (I am working on adding this design now), but I think users should interact pretty much exclusively with the `Package` type. I'll ping you (and anyone else) to chat about what this might look like :+1: